meorg_client/client.py

```diff
@@ -59,28 +70,37 @@ def _make_request(self, method: str, endpoint: str, data: dict=None, headers: di
         # Cast to uppercase, for consisency
         method = method.upper()
 
-        if method not in VALID_METHODS:
-            raise Exception(f'Invalid method {method}')
+        if method not in mcc.VALID_METHODS:
+            raise Exception(f"Invalid method {method}")
 
         # GET/PUT requests have the data interpolated into the url
-        if method in ['GET', 'PUT']:
-            endpoint = endpoint.format(data)
+        if method in mcc.INTERPOLATING_METHODS:
+            endpoint = endpoint.format(**data)
 
-        url = f"{self.base_url}/{endpoint}"
+        url = urljoin(self.base_url, endpoint)
         all_headers = {**self.headers, **headers} if headers else self.headers
-        response = requests.request(method, url, data=data, headers=all_headers)
 
-        if response.status_code == 200:
+        # Update the client with the last response
+        self.last_response = requests.request(
+            method, url, data=data, headers=all_headers
+        )
 
+        # RFC 2616 states status in the 2xx range are considered successful
+        if self.last_response.status_code in mcc.HTTP_STATUS_SUCCESS_RANGE:
             # Return JSON if that's what it is (this should be the default)
-            if response.headers.get('Content-Type', str) == 'application/json':
-                return response.json()
-            
-            return response
+            if (
+                self.last_response.headers.get("Content-Type", str)
+                == mcc.HTTP_CONTENT_TYPES["json"]
+                and return_json == True
+            ):
+                return self.last_response.json()
 
-        else:
-            raise Exception(f"Request failed with status code {response.status_code}: {response.text}")
+            return self.last_response
 
+        else:
+            raise RequestException(
+                self.last_response.status_code, self.last_response.text
+            )
 
```

Comment:
It might be a good idea to create separate methods for each request method. Like make_post_request, make_get_request etc. That might scale better as APIs grow as different methods will have different ways of parsing request and response which might result in a lot of conditions if used in the same method.

Reply:
Done, coming in forthcoming commit. Private methods implemented.

Reply:
Actually, I don't think this is actually necessary as I'd need a function for each of the 5 HTTP methods which do little more than delegate to the `_make_request()` function. This message passing / middle-man behaviour is typically considered poor practice (https://refactoring.guru/smells/middle-man) and it is less redundant to just have this as an argument as it currently is. So I think we'll leave this as is for now.

The API for this is not likely to grow much further than the handful of endpoints we currently have. If we reach the point that necessitates private dedicated methods please feel free to reference this issue and say "I told you so" and I will implement them then.

Reply:
Usually I have seen more than just delegation to another function. Each method could have different headers, different ways to make request and handle response. But if the API have only a few endpoints, then you might not need it.